File: ThreadLab2Java/Threads/src/ArrayMinimumFinder.java
public class ArrayMinimumFinder {
    private final ArrayMinimumManager manager;
    private final int threadCount;

    private final Object lockerForCount = new Object();
    private int completedThreadCount = 0;

    public ArrayMinimumFinder(ArrayMinimumManager manager, int threadCount) {
        this.manager = manager;
        this.threadCount = threadCount;
    }

    public void findMinInArr(int[] arr, int start, int end) {
        int min = arr[start];
        for (int i = start + 1; i < end; i++) {
            if (arr[i] < min) {
                min = arr[i];
            }
        }

        manager.updateMin(min);
    }

    public void startFindingMin(int[] arr) {
        int size = arr.length;
        completedThreadCount = 0;

        long startTime = System.currentTimeMillis();

        for (int i = 0; i < threadCount; i++) {
            Bound bound = getSegment(i, threadCount, size);
            new Thread(() -> threadWorker(arr, bound.start(), bound.end())).start();
        }

        synchronized (lockerForCount) {
            while (completedThreadCount < threadCount) {
                try {
                    lockerForCount.wait();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        }

        long endTime = System.currentTimeMillis();
        System.out.println("---Search time with Synchronized Threads: " + (endTime - startTime) + " ms");
    }

    private void threadWorker(int[] arr, int start, int end) {
        findMinInArr(arr, start, end);
        incThreadCount();
    }

    private void incThreadCount() {
        synchronized (lockerForCount) {
            completedThreadCount++;
            lockerForCount.notify();
        }
    }

    private static Bound getSegment(int index, int arrCount, int length) {
        int segmentLength = length / arrCount;
        int firstIndex = segmentLength * index;
        
        int lastIndex = (index == arrCount - 1) ? length : segmentLength * (index + 1);

        return new Bound(firstIndex, lastIndex);
    }
}

File: ThreadLab2C#/MinimumFinder/ArrayMinimumFinder.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading;

namespace ConsoleApp1
{
    public class ArrayMinimumFinder
    {
        private readonly ArrayMinimumManager _manager;
        private readonly int _threadCount;

        private readonly object _countLocker = new object();
        private int _completedThreadCount = 0;

        public ArrayMinimumFinder(ArrayMinimumManager manager, int threadCount)
        {
            _manager = manager;
            _threadCount = threadCount;
        }

        public void FindMinInArr(int[] arr, int start, int end)
        {
            int min = arr[start];
            for (int i = start + 1; i < end; i++)
            {
                if (arr[i] < min)
                {
                    min = arr[i];
                }
            }

            _manager.UpdateMin(min);
        }

        public void StartFindingMin(int[] arr)
        {
            int size = arr.Length;
            _completedThreadCount = 0;

            var sw = Stopwatch.StartNew();

            for (int i = 0; i < _threadCount; i++)
            {
                var bound = GetSegment(i, _threadCount, size);
                new Thread(() => ThreadWorker(arr, bound.Start, bound.End)).Start();
            }

            lock (_countLocker)
            {
                while (_completedThreadCount < _threadCount)
                {
                    Monitor.Wait(_countLocker);
                }
            }

            sw.Stop();
            Console.WriteLine($"----Search time with Threads: {sw.ElapsedMilliseconds} ms");
        }

        private void ThreadWorker(int[] arr, int start, int end)
        {
            FindMinInArr(arr, start, end);
            IncThreadCount();
        }

        private void IncThreadCount()
        {
            lock (_countLocker)
            {
                _completedThreadCount++;
                Monitor.Pulse(_countLocker);
            }
        }

        private static Bound GetSegment(int index, int arrCount, int length)
        {
            int segmentSize = length / arrCount;
            int start = index * segmentSize;
            int end = (index == arrCount - 1) ? length : start + segmentSize;

            return new Bound(start, end);
        }
    }
}
